### DISCUSSION TITLE
Eliminating `<T>`, step by step.

### DISCUSSION THREAD

**Discussion by longfin** [planetarium/libplanet#2440] [Ideas]
In many cases dealing with `Block<T>` or `Transaction<T>`, we don't need to know the specific type of actions (i.e., `T`) that the block contains since `T` is needed only in limited cases. (i.e., calculating the next states)

So it has been proposed to get rid of this 'T' for a long time too. But since `Block<T>` and `Transaction<T>` had already been used too much, we couldn't get our hands on them easily.

To make this easier, I suggest the following sequence:

1. Instead of excluding `T` from `Block<T>`, `Transaction<T>`, define new interfaces that guarantee only other necessary information (e.g., header).
2. For classes using `Block<T>`, and `Transaction<T>` (e.g., `Swarm<T>`), use the interface used in 1 and remove the `T`.
     - In this case, it is desirable to work from a class (terminal) that has no dependencies as much as possible.
3. Remove `T` from `Block<T>`, and `Transaction<T>`, and only explicitly assume this in classes that need it, such as `ActionEvaluator`.

**Comment by longfin** [comment]
(Note for future workings)

To elide `T` from `IBlockPolicy<T>`, these types there may need to be widened.

- `Transaction<T>` -> `ITransaction`
- `Block<T>` -> `IBlock`
- `BlockChain<T>` -> `IBlockChainStates`

Also, since `ITransaction` and `.BlockAction` will become to agnostic against actual types, we need to provide context info (i.e., `ActionTypeLoaderContext`) to activate `IAction`s.

**Comment by riemannulus** [comment]
https://github.com/planetarium/libplanet/pull/3149

**Comment by ilgyu-yi** [comment]
Closing, since `<T>` has been eliminated from `Block`, `Transaction` and classes using them.
- https://github.com/planetarium/libplanet/pull/3121
- https://github.com/planetarium/libplanet/pull/3123
- https://github.com/planetarium/libplanet/pull/3163
- https://github.com/planetarium/libplanet/pull/3182
- https://github.com/planetarium/libplanet/pull/3183
- https://github.com/planetarium/libplanet/pull/3184
- https://github.com/planetarium/libplanet/pull/3185
- https://github.com/planetarium/libplanet/pull/3193
